tachyon/zk/air/constraint_system/constraint_system.h

```diff
@@ -0,0 +1,95 @@
+#ifndef TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+#define TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+
+#include <stddef.h>
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "third_party/eigen3/Eigen/Core"
+
+#include "tachyon/math/matrix/matrix_types.h"
+#include "tachyon/zk/air/constraint_system/variable.h"
+#include "tachyon/zk/air/expressions/air_evaluator.h"
+
+namespace tachyon::zk::air {
+
+template <typename F>
+class ConstraintSystem {
+ public:
+  ConstraintSystem() = default;
+  ConstraintSystem(size_t preprocessed_width, size_t main_width,
+                   size_t num_public_values)
+      : preprocessed_width_(preprocessed_width),
+        main_width_(main_width),
+        num_public_values_(num_public_values) {}
+
+  void EnforceFirstRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::FirstRow(std::move(constraint)));
+  }
+
+  void EnforceLastRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::LastRow(std::move(constraint)));
+  }
+
+  void EnforceTransitionConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::Transition(std::move(constraint)));
+  }
+
+  void EnforceConstraint(std::unique_ptr<Expression<F>> constraint) {
+    constraints_.push_back(std::move(constraint));
+  }
+
+  bool IsSatisfied(AirEvaluator<F>& evaluator,
+                   const std::vector<F>& public_values,
+                   std::unique_ptr<const math::RowMajorMatrix<F>> preprocessed,
+                   const math::RowMajorMatrix<F>& main) const {
+    CHECK_EQ(public_values.size(), num_public_values_);
+    CHECK_EQ(main.cols(), main_width_);
+    CHECK(preprocessed != nullptr || preprocessed_width_ == 0);
+    if (preprocessed) {
+      CHECK_EQ(preprocessed->cols(), preprocessed_width_);
+      CHECK_EQ(preprocessed->rows(), main.rows());
+    }
+
+    evaluator.set_public_values(public_values);
+    evaluator.set_num_rows(main.rows());
+
+    for (Eigen::Index i = 0; i < main.rows() - 1; ++i) {
+      const math::RowMajorMatrix<F>& main_window = main.middleRows(i, 2);
+
+      if (preprocessed) {
+        const math::RowMajorMatrix<F>& prep_window =
+            preprocessed->middleRows(i, 2);
+
+        evaluator.SetCurrentWindowData(i, &prep_window, &main_window);
+
+        for (const std::unique_ptr<Expression<F>>& constraint : constraints_) {
+          if (!evaluator.Evaluate(constraint.get()).IsZero()) {
+            return false;
+          }
+        }
+      } else {
+        evaluator.SetCurrentWindowData(i, nullptr, &main_window);
+
+        for (const std::unique_ptr<Expression<F>>& constraint : constraints_) {
+          if (!evaluator.Evaluate(constraint.get()).IsZero()) {
+            return false;
+          }
+        }
+      }
```

Comment:
```suggestion
      const math::RowMajorMatrix<F>& main_window = main.middleRows(i, 2);
      const math::RowMajorMatrix<F>& prep_window =
          preprocessed? preprocessed->middleRows(i, 2): nullptr;

      evaluator.SetCurrentWindowData(i, &prep_window, &main_window);

      for (const std::unique_ptr<Expression<F>>& constraint : constraints_) {
        if (!evaluator.Evaluate(constraint.get()).IsZero()) {
          return false;
        }
      }
```

Reply:
I don't think this would compile since you can't assign nullptr to reference.

```
const math::RowMajorMatrix<F>* prep_window =
	          preprocessed ? &preprocessed->middleRows(i, 2): nullptr;
```
	          
would not work as well since object will be out of scope.

Reply:
Oh, you're right. What I tried to say was that if possible, please reduce code duplication!

Reply:
I think it's hard since `Evaluate` must be called within the `if (preprocessed) {}` scope or `prep_window` will go out of scope and result UB.

Reply:
Since it's minor, i'll resolve.